blib/bindoc/GATK.pl.bak.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GATK.PL.BAK 1"
.TH GATK.PL.BAK 1 "2015-07-27" "perl v5.10.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
GATK.pl
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fB\s-1GATK\s0.pl\fR [options] <\-\-config> <\-\-run_id> <\-\-output_dir>
.PP
.Vb 4
\&        Arguments:
\&        \-\-config                                A YAML file containing BAM files and sample names for processing
\&        \-\-run_id                                A name/ID for this GATK run, suggest to use the project, dataset or cohort name
\&        \-\-output_dir                            Output directory (for example: /scratch2/user/your_user_name).
\&        
\&        Options:
\&        \-\-help                                  Brief help message
\&        \-\-man                                   Full documentation
\&        \-\-gatk_version                          Version of the GATK program, e.g. "3.3.0", "2.4.9", etc.
\&        \-\-ref                                   Path for reference genome fasta file (default:hg19)
\&        \-\-indels                                Insertion/deletion file in VCF format
\&        \-\-dbsnp                                 dbSNP file in VCF format
\&        \-\-ref_type                              Reference genome (default:hg19) Other acceptable reference genome is b37 from the broad institute
\&        \-\-indel_realigner_target_creator        Perform RealingerTargetCreator (default:Y)
\&        \-\-indel_realigner                       Perform IndelReaglinger (default:Y)
\&        \-\-create_recalibration_table            Perform CreateRecalibrationTable (default:Y)
\&        \-\-print_reads                           Perfrom PrintReads (default:Y)
\&        \-\-samtools_reheader                     Perform samtools reheader (default:Y)
\&        \-\-Pre_processing                        Perform BAM processing (default:Y)
\&        \-\-SNP_calling                           Perform SNP calling (default:Y)
\&        \-\-merge_raw_vcf                         Perfrom MergeRawVcf (default:Y)
\&        \-\-variant_recalibrator                  Perfrom VariantRecalibrator (default:Y)
\&        \-\-apply_recalibration                   Perfrom ApplyRecalibration (default:Y)
\&
\&        Examples:
\&        
\&        ### Simple case, assuming everying will run smoothly and the default values are fine to you:
\&        perl GATK.pl \-\-config tumour_normal_paried.yaml \-\-output_dir /scratch2/user/jwang/TEST \-\-gatk_version 3.3.0 \-\-variant_caller HC \-\-run_id POPCURE
\&        
\&        ### Assume the pipeline failed at the step of indexing recalibrated and reheadered bams due to insuficient memory
\&        ### for some high read\-depth/coverage bams. You have edited the corresponding script files to increase
\&        ### memory requirement. You can resume your run by this command:
\&        perl GATK.pl \-\-config tumour_only.yaml \-\-output_dir /scratch2/user/jwang/TEST \e
\&                \-\-gatk_version 2.4.9 \-\-variant_caller UG \-\-run_id CPC_GENE
\&                \-\-indel_realigner_target_creator N  \-\-indel_realigner N \e
\&                \-\-create_recalibration_table N
\&        
\&        ### Assume some jobs failed at the stage of variant calling. You identified the cause
\&        ### and fixed the problem. Now you can resume the run and skipping the GATK phase I runs:
\&        perl GATK.pl \-\-config tumour_only.yaml \-\-output_dir /scratch2/user/jwang/TEST \e
\&                \-\-gatk_version 3.3.0 \-\-variant_caller UG \-\-run_id CPC_GENE \-\-Pre_processing N
.Ve
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-\-help\fR" 8
.IX Item "--help"
Print a brief help message and exit.
.IP "\fB\-\-man\fR" 8
.IX Item "--man"
Print the manual page.
.IP "\fB\-\-config\fR" 8
.IX Item "--config"
Config \s-1YAML\s0 file containing samples with corresponding \s-1BAM\s0 file locations
.IP "\fB\-\-sample\fR" 8
.IX Item "--sample"
Name for a job of the \s-1GATK\s0 pipeline
.IP "\fB\-\-dir\fR" 8
.IX Item "--dir"
Output directory name. /scratch/user/ is recommended because this \s-1GATK\s0 script can generated a few \s-1TB\s0 of intermediate files if your sample is whole gneome sequencing data.
.IP "\fB\-\-ref\fR" 8
.IX Item "--ref"
Path for reference genome fasta file. Default is the path for the b37_decoy reference genome.
.IP "\fB\-\-ref_type\fR" 8
.IX Item "--ref_type"
Reference genome (default:b37_decoy, from Broad Institute, used in 1kg phaseII)
.IP "\fB\-\-pre_processing\fR" 8
.IX Item "--pre_processing"
Perform \s-1BAM\s0 processing including RealignerTargetCreator, IndelRealigner, BaseRecalibrator, PrintReads, and merging and indexing \s-1BAM\s0 files using Picard.
(default:Y)
.IP "\fB\-\-SNP_calling\fR" 8
.IX Item "--SNP_calling"
Perform \s-1SNP\s0 calling including HaplotypeCaller or UnifiedGenotyper, VariantRecalibrator, ApplyRecalibration, and merging \s-1VCF\s0 files. (default:Y)
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fB\s-1GATK\s0.pl\fR Run the Genome Analysis Toolkit pipeline.
.SH "AUTHOR"
.IX Header "AUTHOR"
Jianxin Wang
.PP
Center for Computational Research, University at Buffalo
.SH "SEE ALSO"
.IX Header "SEE ALSO"


Changes
Revision history for NGS-Tools-GATK

0.01    2015-05-20
        First version
0.05    2016-10-20, updated to use GATK v3.6, skipped indel realignment


blib/script/GATK.pl
#!/usr/bin/env perl

### GATK.pl ##############################################################################
# Run the Genome Analysis Toolkit pipeline.

### INCLUDES ######################################################################################
use warnings;
use strict;
use Carp;
use Getopt::Long;
use Pod::Usage;
use Path::Class;
use File::Path;
use File::Spec;
use File::Basename;
use YAML qw(LoadFile);

use NGS::Tools::GATK;

### COMMAND LINE DEFAULT ARGUMENTS ################################################################
our %opts = (
    debug                      => 'N',
    gatk_version               => '3.8.0',  # set default version
    phaseI                     => 'Y',
    phaseII                    => 'Y',
    phaseIII                   => 'Y',
    create_recalibration_table => 'Y',
    create_recalibrated_bam    => 'Y',
    split_bam_by_chr           => 'Y',
    run_haplotype_caller       => 'Y',
    joint_calling_genotypes    => 'Y',
    cat_variants               => 'Y',
    variant_recalibrator       => 'Y',
    apply_recalibration        => 'Y',
    ref                        => '/gpfs/scratch/jw24/GATK/gatk_bundle/2.8/human_g1k_v37_decoy.fasta',
    dbsnp                      => '/gpfs/scratch/jw24/GATK/gatk_bundle/2.8/dbsnp_138.b37.vcf',
    indels                     => '/gpfs/scratch/jw24/GATK/gatk_bundle/2.8/dbsnp_138.b37.vcf',
    hapmap                     => '/gpfs/scratch/jw24/GATK/gatk_bundle/2.8/hapmap_3.3.b37.vcf',
    omni                       => '/gpfs/scratch/jw24/GATK/gatk_bundle/2.8/1000G_omni2.5.b37.vcf',
    g1k_snp                    => '/gpfs/scratch/jw24/GATK/gatk_bundle/2.8/1000G_phase1.snps.high_confidence.b37.vcf',
    g1k_indel                  => '/gpfs/scratch/jw24/GATK/gatk_bundle/2.8/1000G_phase1.indels.b37.vcf',
    mills                      => '/gpfs/scratch/jw24/GATK/gatk_bundle/2.8/Mills_and_1000G_gold_standard.indels.b37.vcf'
#    modules                    => 'samtools,java/1.8.0_45,vcftools,picard/2.2.1,R/3.2.0'
);

GetOptions(
    \%opts,
    "help|?",
    "man",
    "gatk_version|v:s" => \$opts{'gatk_version'},
    "run_id:s",
    "config|c:s",
    "output_dir:s",
    "data_type:s",
    "debug:s"                      => \$opts{'debug'},
    "ref|r:s"                      => \$opts{'ref'},
    "dbsnp:s"                      => \$opts{'dbsnp'},
    "indels:s"                     => \$opts{'indels'},
    "hapmap:s"                     => \$opts{'hapmap'},
    "omni:s"                       => \$opts{'omni'},
    "g1k_snp:s"                    => \$opts{'g1k_snp'},
    "g1k_indel:s"                  => \$opts{'g1k_indel'},
    "mills"                        => \$opts{'mills'},
    "modules:s"                    => \$opts{'modules'},
    "phaseI"                       => \$opts{'phaseI'},
    "phaseII"                      => \$opts{'phaseII'},
    "phaseIII"                     => \$opts{'phaseIII'},
    "create_recalibration_table:s" => \$opts{'create_recalibration_table'},
    "create_recalibrated_bam:s"    => \$opts{'create_recalibrated_bam'},
    "split_bam_by_chr"             => \$opts{'split_bam_by_chr'},
    "run_haplotype_caller:s"       => \$opts{'run_haplotype_caller'},
    "joint_calling_genotypes:s"    => \$opts{'joint_calling_genotypes'},
    "cat_variants:s"               => => \$opts{'cat_variants'},
    "variant_recalibrator:s"       => \$opts{'variant_recalibrator'},
    "apply_recalibration:s"        => \$opts{'apply_recalibration'}
) or pod2usage(2);
GetOptions();

####################################################################################################
# Check commandline sanity
####################################################################################################
if ( !$opts{'config'} or !$opts{'output_dir'} or !$opts{'run_id'} or !$opts{'data_type'}) {
    pod2usage(1);
}

if ( $opts{'help'} ) { pod2usage(1) }
if ( $opts{'man'} ) { pod2usage( -exitstatus => 0, -verbose => 2 ) }

# Load the config yaml file
my $config;
if ( $opts{'config'} ) {
    $config = LoadFile( $opts{'config'} );
}
else {
    croak "Config yaml file not found. Please check your commandline\n";
}

####################################################################################################
# Root directories to store output files
####################################################################################################
my $gatk_version  = $opts{'gatk_version'};
my $parent_dir    = File::Spec->rel2abs( $opts{'output_dir'} );
my $phaseI_root   = $parent_dir . '/phaseI_' . $gatk_version;
my $phaseII_root  = $parent_dir . '/phaseII_' . $gatk_version;
my $phaseIII_root = $parent_dir . '/phaseIII_' . $gatk_version;

my $gatk = NGS::Tools::GATK->new(
    bam           => $config,
    config        => $config,
    opts          => \%opts,
    phaseI_root   => $phaseI_root,
    phaseII_root  => $phaseII_root,
    phaseIII_root => $phaseIII_root,
);

####################################################################################################
# Prepare directories for output files
####################################################################################################
# Switch to working directory
chdir($parent_dir);

# Initialize directory structure
$gatk->dir_structure();

foreach my $phasedir ( 'phaseI', 'phaseII', 'phaseIII' ) {
    my $phaseroot = $parent_dir . '/' . $phasedir . '_' . $gatk_version;
    File::Path::make_path( File::Spec->rel2abs($phaseroot) );
    chdir( File::Spec->rel2abs($phaseroot) ) or croak "$!";

    my @phase_dirs = values %{ $gatk->dir->{$phasedir} };
    File::Path::make_path(@phase_dirs);

    # Switch back to parent dir
    chdir( File::Spec->rel2abs($parent_dir) ) or croak "$!";
}

# Fill in the sample_names to the $gatk object
my $sample_names = $gatk->get_sample_name_from_bam();

####################################################################################################
# GATK phase I, indel realigner and base quality recalibration
####################################################################################################

my $base_recalibration = $gatk->run_base_recalibration();

my $recalibrated_bam = $base_recalibration->create_recalibrated_bams();

####################################################################################################
# GATK phase II, variant calling
####################################################################################################

# make variant calls with HaplotypeCaller
# due to long running time and high memory requirement, split recalibrated bams by chr
my $split_bam = $recalibrated_bam->split_bam_by_chr();

my $htc = $split_bam->run_haplotype_caller();

my $genotyped_vcf = $htc->joint_calling_genotypes();

my $combined_variants = $genotyped_vcf->run_cat_variants();

####################################################################################################
# GATK phaseIII, variant recalibration
####################################################################################################

my $snp_recalibration              = $combined_variants->run_variant_recalibrator('snp');
my $apply_variant_recalibrator_snp = $snp_recalibration->run_apply_recalibration('snp');

my $indel_recalibration               = $apply_variant_recalibrator_snp->run_variant_recalibrator('indel');
my $apply_variant_recalibration_indel = $indel_recalibration->run_apply_recalibration('indel');

my $merge_bams = $apply_variant_recalibration_indel->merge_hc_bams();


# if we reach here, everything is ok
print "All jobs submitted...\n";

exit;

__END__


=head1 NAME

GATK.pl

=head1 SYNOPSIS

B<GATK.pl> [options] <--config> <--run_id> <--output_dir>

	Arguments:
	--config				A YAML file containing BAM files and sample names for processing
	--run_id				A name/ID for this GATK run, suggest to use the project, dataset or cohort name
	--output_dir			Output directory (for example: /scratch2/user/your_user_name).
    --data_type             Indicate if the input sequence data are from whole genome sequencing or whole exome sequencying. valid values are "wgs" or "wes"
	
	Options:
	--help					Brief help message
	--man					Full documentation
	--gatk_version			Version of the GATK program, e.g. "3.3.0", "2.4.9", etc. default "3.6.0"
	--ref					Path for reference genome fasta file (default:hg19)
	--indels				Insertion/deletion file in VCF format
	--dbsnp					dbSNP file in VCF format
	--ref_type				Reference genome (default:hg19) Other acceptable reference genome is b37 from the broad institute
	--indel_realigner_target_creator	Perform RealingerTargetCreator (default:Y)
	--indel_realigner			Perform IndelReaglinger (default:Y)
	--create_recalibration_table		Perform CreateRecalibrationTable (default:Y)
	--merge_raw_vcf				Perfrom MergeRawVcf (default:Y)
	--variant_recalibrator			Perfrom VariantRecalibrator (default:Y)
	--apply_recalibration			Perfrom ApplyRecalibration (default:Y)

	Examples:
	
	### Simple case, assuming everying will run smoothly and the default values are fine to you:
	perl gatk.pl --config tumour_normal_paried.yaml --output_dir /scratch2/user/jwang/TEST --gatk_version 3.5.0 --data_type wgs --run_id SIMWGS
	
	
=head1 OPTIONS

=over 8

=item B<--help>

Print a brief help message and exit.

=item B<--man>

Print the manual page.

=item B<--config>

Config YAML file containing samples with corresponding BAM file locations

=item B<--sample>

Name for a job of the GATK pipeline

=item B<--dir>

Output directory name. /scratch/user/ is recommended because this GATK script can generated a few TB of intermediate files if your sample is whole gneome sequencing data.

=item B<--ref>

Path for reference genome fasta file. Default is the path for the b37_decoy reference genome.

=item B<--ref_type>

Reference genome (default:b37_decoy, from Broad Institute, used in 1kg phaseII)

=item B<--pre_processing>

Perform BAM processing including RealignerTargetCreator, IndelRealigner, BaseRecalibrator, PrintReads, and merging and indexing BAM files using Picard.
(default:Y)

=item B<--SNP_calling>

Perform SNP calling including HaplotypeCaller or UnifiedGenotyper, VariantRecalibrator, ApplyRecalibration, and merging VCF files. (default:Y)

=back

=head1 DESCRIPTION

B<GATK.pl> Run the Genome Analysis Toolkit pipeline.

=head1 AUTHOR

Jianxin Wang

Center for Computational Research, University at Buffalo

=head1 SEE ALSO

=cut

